### PR TITLE
Replace TacticWorldParamsUpdateVisitor with an override method in each tactic

### LINKED ISSUE #1710: Replace TacticWorldParamsUpdateVisitor with `Tactic::updateWorldParams`
### Description of the task
Replace `TacticWorldParamsUpdateVisitor` with `Tactic::updateWorldParams(World world)`, which each Tactic will override. This will simplify the code, because the complexity involved with `TacticWorldParamsUpdateVisitor` doesn't add much. We can also make the `MutableTacticVisitor` nonmutable and call it `TacticVisitor` instead

### Acceptance criteria

- [ ] Replace TacticWorldParamsUpdateVisitor with `Tactic::updateWorldParams`
- [ ] Clean up mutable tactic visitor

### PR DESCRIPTION
**Pull request by PhilD71**
### Description

Removed the visitor design for updating the world of every tactic, done by replacing the TacticWorldsParamUpdateVisitor with an an updateWorldParams(world world) method override in each tactic

### Testing Done

Passes all bazel tests in stp, where the tactic visitor was used
### Resolved Issues

resolves #1710 
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

Still a little unclear on how to make the Mutable Tactic Visitor immutable.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.h [hunk 1]
```diff
@@ -18,6 +18,13 @@ class StopTestTactic : public Tactic
      */
     explicit StopTestTactic(bool loop_forever = false);
 
+    /**
+     * Updates the world parameters for this ChipTactic.
+     *
+     * @param World The current state of the world
+     */
+    void updateWorldParams(const World& world) override;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.h]: it's not necessary to have a function comment since we're overriding

**Review by jonathanlew — CHANGES_REQUESTED** [review]
on the right track, just a few comments

File: src/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp [hunk 2]
```diff
@@ -12,6 +12,8 @@ CherryPickTactic::CherryPickTactic(const World& world, const Rectangle& target_r
     pass_generator.setTargetRegion(target_region);
 }
 
+
```
Comment (EvanMorcom) [inline src/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp]: Nit:

Looks like these are unnecessary

File: src/software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.h [hunk 3]
```diff
@@ -27,16 +27,23 @@ class DefenseShadowEnemyTactic : public Tactic
...
     /**
      * Updates the world parameters for this Tactic.
      *
      * @param field The current state of the field
-     * @param friendly_team The current state of the friendly team
-     * @param enemy_team The current state of the enemy team
-     * @param ball The current state of the ball
+     * @param updated_friendly_team The current state of the friendly team
+     * @param updated_enemy_team The current state of the enemy team
+     * @param updated_ball The current state of the updated_ball
```
Comment (EvanMorcom) [inline src/software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.h]: Nit:

"current state of the _ball_'

**Comment by PhilD71** [comment]
Made changes to the overrides, added changes to the tests as well. I finally made `MutableTacticVisitor` immutable by making all of its methods have a constant parameter, so it becomes immutable. Renamed it to `TacticVisitor`.

**Review by jonathanlew — COMMENTED** [review]
lgtm pending other reviewers' comments

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm :100: nice work, we should probably also get rid of the action world param update visitor, ill make a ticket for that

File: src/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp]: ^ bumping this comment

File: src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/test_tactics/stop_test_tactic.h]: ^bumping this

File: src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.h [hunk 4]
```diff
@@ -18,6 +18,13 @@ class MoveTestTactic : public Tactic
      */
     explicit MoveTestTactic(bool loop_forever = false);
 
+    /**
+     * Updates the world parameters for this ChipTactic.
+     *
+     * @param World The current state of the world
+     */
+    void updateWorldParams(const World& world) override;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.h:21]: it's not necessary to have a function comment since we're overriding

**Review by jonathanlew — COMMENTED** [review]
still a few remaining nits

File: src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.h (hunk 4 above)
Comment (PhilD71) [reply src/software/ai/hl/stp/tactic/test_tactics/move_test_tactic.h:21]: resolved

File: src/software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.h (hunk 3 above)
Comment (PhilD71) [reply src/software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.h]: resolved

**Comment by PhilD71** [comment]
> still a few remaining nits

resolved!

**Review by jonathanlew — APPROVED** [review]
LGTM! :+1:

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**